storage/http_client.go

```diff
@@ -684,7 +684,56 @@ func (c *httpStorageClient) ComposeObject(ctx context.Context, req *composeObjec
 	return newObject(obj), nil
 }
 func (c *httpStorageClient) RewriteObject(ctx context.Context, req *rewriteObjectRequest, opts ...storageOption) (*rewriteObjectResponse, error) {
-	return nil, errMethodNotSupported
+	s := callSettings(c.settings, opts...)
+	rawObject := req.attrs.toRawObject("")
+	call := c.raw.Objects.Rewrite(req.srcBucket, req.srcObject, req.dstBucket, req.dstObject, rawObject)
+
+	call.Context(ctx).Projection("full")
+	if req.token != "" {
+		call.RewriteToken(req.token)
+	}
+	if req.dstKeyName != "" {
+		call.DestinationKmsKeyName(req.dstKeyName)
+	}
+	if req.predefinedACL != "" {
+		call.DestinationPredefinedAcl(req.predefinedACL)
+	}
+	if err := applyConds("Copy destination", req.gen, req.conds, call); err != nil {
+		return nil, err
+	}
+	if err := applySourceConds(req.srcGen, req.srcConds, call); err != nil {
+		return nil, err
+	}
+	// The userProject, whether source or destination project, is decided by the code calling the interface.
+	if s.userProject != "" {
+		call.UserProject(s.userProject)
+	}
+	// Set destination encryption headers.
+	if err := setEncryptionHeaders(call.Header(), req.dstEncryptionKey, false); err != nil {
+		return nil, err
+	}
+	// Set source encryption headers.
+	if err := setEncryptionHeaders(call.Header(), req.srcEncryptionKey, true); err != nil {
+		return nil, err
+	}
+	var res *raw.RewriteResponse
+	var err error
+	setClientHeader(call.Header())
+
+	retryCall := func() error { res, err = call.Do(); return err }
+
+	if err := run(ctx, retryCall, s.retry, s.idempotent, setRetryHeaderHTTP(call)); err != nil {
+		return nil, err
+	}
+
+	r := &rewriteObjectResponse{
```

Comment:
https://github.com/googleapis/google-cloud-go/blob/0c820891a987f4a1c6563106e4e54b8e7b46cb30/storage/copy.go#L100 

... requires ObjectSize as well; is there a way we can pass this back also? On the other hand, it doesn't look like `token` is used.

Reply:
That is reachable by doing `res.resource.ObjectSize` of a rewriteObjectResponse. 

Token is necessary because currently `callRewrite` sets `c.Token`. We will need to do the same thing with the code that handles the `rewriteObjectResponse` in copy.go

Reply:
Makes sense, I wasn't looking at the response closely enough.